source/AI.cpp

```diff
@@ -408,17 +431,22 @@ void AI::UpdateKeys(PlayerInfo &player, Command &activeCommands)
 	// Get rid of any invalid orders. Carried ships will retain orders in case they are deployed.
 	for(auto it = orders.begin(); it != orders.end(); )
 	{
-		if(it->second.type & Orders::REQUIRES_TARGET)
+		if(it->second.type == Orders::MINING && it->first->Cargo().Free() && !(it->second.targetAsteroid.lock()))
+		{
+			orders[it->first].type = Orders::HARVEST;
+		}
+		else if(it->second.type & Orders::REQUIRES_TARGET)
 		{
 			shared_ptr<Ship> ship = it->second.target.lock();
+			shared_ptr<Minable> asteroid = it->second.targetAsteroid.lock();
 			// Check if the target ship itself is targetable.
 			bool invalidTarget = !ship || !ship->IsTargetable() || (ship->IsDisabled() && it->second.type == Orders::ATTACK);
 			// Check if the target ship is in a system where we can target.
 			// This check only checks for undocked ships (that have a current system).
 			bool targetOutOfReach = !ship || (it->first->GetSystem() && ship->GetSystem() != it->first->GetSystem()
 					&& ship->GetSystem() != flagship->GetSystem());
 
-			if(invalidTarget || targetOutOfReach)
+			if(!asteroid && (invalidTarget || targetOutOfReach))
```

Comment:
It is not possible to update just `invalidTarget`.  I must update both `invalidTarget` and `targetOutOfReach`.  That's the reason why I moved `!asteroid` into this conditional because the logic looks much more simple and fewer conditional checks.

Reply:
I think I found a cleaner way to update this; you'll see on push